File: server/basketball/services/prueba_antropometrica_service.py
"""
Servicio para la gestión de Pruebas Antropométricas
"""

import logging
from typing import List, Optional, Dict, Any
from django.core.exceptions import ValidationError
from ..dao.prueba_antropometrica_dao import PruebaAntropometricaDAO
from ..models import PruebaAntropometrica

logger = logging.getLogger(__name__)


class PruebaAntropometricaService:
    """Servicio que encapsula la lógica de negocio para Pruebas Antropométricas"""
    
    def __init__(self):
        self.dao = PruebaAntropometricaDAO()
    
    def create_prueba(self, data: Dict[str, Any]) -> PruebaAntropometrica:
        """
        Crea una nueva prueba antropométrica.
        
        Args:
            data: Diccionario con los datos de la prueba
            
        Returns:
            PruebaAntropometrica: Instancia creada
            
        Raises:
            ValidationError: Si los datos no son válidos
        """
        # Validaciones adicionales si son necesarias
        peso = data.get('peso')
        estatura = data.get('estatura')
        
        if peso is not None and float(peso) <= 0:
            raise ValidationError("El peso debe ser mayor a 0")
            
        if estatura is not None and float(estatura) <= 0:
            raise ValidationError("La estatura debe ser mayor a 0")
            
        return self.dao.create(**data)
    
    def update_prueba(self, pk: int, data: Dict[str, Any]) -> Optional[PruebaAntropometrica]:
        """
        Actualiza una prueba existente.
        
        Args:
            pk: ID de la prueba
            data: Datos a actualizar
            
        Returns:
            PruebaAntropometrica | None: Instancia actualizada
        """
        return self.dao.update(pk, **data)
    
    def delete_prueba(self, pk: int) -> bool:
        """
        Elimina (soft delete) una prueba.
        
        Args:
            pk: ID de la prueba
            
        Returns:
            bool: True si se eliminó correctamente
        """
        return self.dao.soft_delete(pk)
    
    def get_prueba_by_id(self, pk: int) -> Optional[PruebaAntropometrica]:
        """Obtiene una prueba por ID"""
        return self.dao.get_by_id(pk)
    
    def get_all_pruebas(self) -> List[PruebaAntropometrica]:
        """Obtiene todas las pruebas activas"""
        return list(self.dao.get_activas())
    
    def get_pruebas_by_atleta(self, atleta_id: int) -> List[PruebaAntropometrica]:
        """Obtiene pruebas de un atleta"""
        return list(self.dao.get_by_atleta(atleta_id))
    
    def search_pruebas(self, search_term: str) -> List[PruebaAntropometrica]:
        """Busca pruebas por término"""
        return list(self.dao.search_pruebas(search_term))

File: client/src/api/pruebaAntropometricaService.js
/**
 * Servicio para gestión de Pruebas Antropométricas
 * Conecta con el endpoint /api/v1/pruebas-antropometricas del backend
 */

import apiClient from './apiClient'
import { ENDPOINTS } from '../config/api.config'

const PruebaAntropometricaService = {
  /**
   * Obtener todas las pruebas antropométricas
   * @param {Object} params - Parámetros de filtro y paginación
   */
  getAll: async (params = {}) => {
    const response = await apiClient.get(ENDPOINTS.PRUEBAS_ANTROPOMETRICAS, { params })
    return response.data
  },

  /**
   * Obtener una prueba por ID
   * @param {number} id - ID de la prueba
   */
  getById: async (id) => {
    const response = await apiClient.get(`${ENDPOINTS.PRUEBAS_ANTROPOMETRICAS}${id}/`)
    return response.data
  },

  /**
   * Crear una nueva prueba antropométrica
   * @param {Object} data - Datos de la prueba
   */
  create: async (data) => {
    const response = await apiClient.post(ENDPOINTS.PRUEBAS_ANTROPOMETRICAS, data)
    return response.data
  },

  /**
   * Actualizar una prueba existente
   * @param {number} id - ID de la prueba
   * @param {Object} data - Datos a actualizar
   */
  update: async (id, data) => {
    const response = await apiClient.put(`${ENDPOINTS.PRUEBAS_ANTROPOMETRICAS}${id}/`, data)
    return response.data
  },

  /**
   * Eliminar una prueba
   * @param {number} id - ID de la prueba
   */
  delete: async (id) => {
    const response = await apiClient.delete(`${ENDPOINTS.PRUEBAS_ANTROPOMETRICAS}${id}/`)
    return response.data
  },

  /**
   * Obtener pruebas por atleta
   * @param {number} atletaId - ID del atleta
   */
  getByAtleta: async (atletaId) => {
    const response = await apiClient.get(`${ENDPOINTS.PRUEBAS_ANTROPOMETRICAS}atleta/${atletaId}/`)
    return response.data
  },
}

export default PruebaAntropometricaService
